Adding proxy options to Pytorch Hub

JubilantJerry opened the pull request:
Implements the feature mentioned in https://github.com/pytorch/pytorch/issues/96412.
Let me know what else needs to be contributed in addition to the code that implements the feature itself, or if there is a more standardized approach toward implementing some of the logic that I added to the Python file.

cc @nairbv @NicolasHug @vmoens @jdsgomes

janeyx99 commented:
@vmoens could you take a look?

JubilantJerry commented:
EasyCLA is now done. I did not know how to do this step back when creating the ticket. So it should be possible to move forward on the issue now.

vmoens commented on torch/hub.py:
```diff
@@ -752,3 +761,17 @@ def load_state_dict_from_url(
     return torch.load(cached_file, map_location=map_location, weights_only=weights_only)
+
+
+def set_proxy(protocol, host, port):
```
Could you add some docstrings here since it's a public method?
Would be nice to have a two line example of how to use it.

JubilantJerry replied:
Added docstrings for the public methods. The example is short enough to fit in one line.

JubilantJerry committed:
Added proxy options to Pytorch Hub

Added a function that changes the function used when Pytorch Hub accesses the internet.
As a result, PyTorch Hub can be configured to use a proxy server.
If PySocks is installed, Pytorch Hub can use SOCKSv4 and SOCKSv5 proxies, in addition to HTTP and HTTPS.

NicolasHug commented:
Thanks for the PR @JubilantJerry 

Instead of supporting a functionality such as `torch.hub.set_proxy()`, shouldn't it be possible for users to manually create a context manager to set the proxy outside of pytorch? E.g. something like

```py
def set_proxy(...):
    # ....

with set_proxy(...):
    torch.hub.load(...)
```

where `set_proxy()` is **not** part of torch core?

It would sound like a more natural solution - if such a helper is hard to implement right now, we can try to make it easier by internally relying on the right utils - although we're only relying on `requests` so that's pretty standard.

We'd prefer going this route instead of introducing a `torch.hub.set_proxy()` util because whether such a proxy tool is in scope is debatable, and also it comes with significant maintenance cost (we'd need new tests that require internet access etc., and this is never easy).

peacefulotter commented:
Hi, may someone take a look at this? IMO, this feature would be valuable
